funcbench/main.go

```diff
@@ -0,0 +1,281 @@
+		}
+		if cmd.ProcessState.ExitCode() != 0 {
+			if err := postComment(clt, owner, repo, prNumber, "Git merge failed"); err != nil {
+				log.Fatalln(err)
+			}
+		}
+
+		data, err = exec.Command("git", "reset").CombinedOutput()
+		if err != nil {
+			log.Fatalln("Error: ", err, "Command out: ", string(data))
+		}
+
+		// Benchmark the with pullrequest changes.
+		if race == "-no-race" {
+			cmd = exec.Command("go", "test", "-bench", fmt.Sprintf("^%s$", issueComment), "-benchmem", "-v", "./...")
+		} else {
+			cmd = exec.Command("go", "test", "-bench", fmt.Sprintf("^%s$", issueComment), "-benchmem", "-race", "-v", "./...")
+		}
+		data, err = cmd.CombinedOutput()
+		log.Println(string(data))
+		if err != nil || cmd.ProcessState.ExitCode() != 0 {
+			if err := postComment(clt, owner, repo, prNumber, fmt.Sprintf("Go test for this pull request failed. %s", logLink)); err != nil {
+				log.Fatalln(err)
+			}
+			log.Fatalln(err)
+		}
+		err = ioutil.WriteFile(newPath, data, 0644)
+		if err != nil {
+			log.Fatalln(err)
+		}
+
+		// Checkout to the comparing branch.
+		err = filepath.Walk(os.Getenv("GITHUB_WORKSPACE"), func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() || strings.Contains(path, ".git/") || strings.HasSuffix(info.Name(), "test.go") {
+				return nil
+			}
+
+			data, err = exec.Command("git", "checkout", "--", path).CombinedOutput()
+			if err != nil {
+				log.Println(string(data))
+
+				// New file appeared and is not tracked, then it can be removed
+				if strings.Contains(string(data), "did not match any file") {
+					data, err = exec.Command("rm", path).CombinedOutput()
+					if err != nil {
+						log.Fatalln("Error: ", err, "Command out: ", string(data))
+					}
+					return nil
+				}
+				log.Fatalln(err)
+			}
+			return nil
+		})
+		if err != nil {
+			log.Fatalln(err)
+		}
+
+		// Benchmark the comparing branch.
+		if race == "-no-race" {
+			cmd = exec.Command("go", "test", "-bench", fmt.Sprintf("^%s$", issueComment), "-benchmem", "-v", "./...")
+		} else {
+			cmd = exec.Command("go", "test", "-bench", fmt.Sprintf("^%s$", issueComment), "-benchmem", "-race", "-v", "./...")
+		}
+		data, err = cmd.CombinedOutput()
+		log.Println(string(data))
+		if err != nil || cmd.ProcessState.ExitCode() != 0 {
+			if err := postComment(clt, owner, repo, prNumber, fmt.Sprintf("Go test on branch %s failed. %s", branch, logLink)); err != nil {
+				log.Fatalln(err)
+			}
+			log.Fatalln(err)
+		}
+		err = ioutil.WriteFile(oldPath, data, 0644)
+		if err != nil {
+			log.Fatalln(err)
+		}
+
+		// Compare the benchmarks.
+		data, err = exec.Command(strings.Join([]string{os.Getenv("GOPATH"), "/bin/benchcmp"}, ""), oldPath, newPath).CombinedOutput()
+		if err != nil {
+			log.Println(string(data))
+			if err := postComment(clt, owner, repo, prNumber, fmt.Sprintf("Error: `benchcmp` failed. %s", logLink)); err != nil {
+				log.Fatalln(err)
+			}
+			log.Fatalln(err)
```

Comment:
you can simplify this by just calling log.Fatal once.

Reply:
Wouldn't the error be overwritten by the `postComment` function?

Reply:
yeah it is a bit confusing as now if `postComment` fails you will never see the `exec.Command` error as  log.Fatalln(err) exits the program immediately

I would say there is too much logic in the main func. Lets brake each step into smaller funcs and  do some proper error handling and wrapping. `errors.Wrap`